### PR TITLE
Feature/resizing from all corners

### PR DESCRIPTION
**Pull request by minrows**
### Checklist
- [x] I added multiple screenshots/screencasts of my UI changes

### Motivation and Context
This feature enables resizing of an elements from all 4 corner.
Issue Number: https://github.com/ls1intum/Apollon/issues/71

### Description
Previously resizing of an element could be done only from the bottom right corner.
This feature will now enable users to resize it from all 4 corners.

### Steps for Testing
1. Open up Apollon and add any Elements
2. Try resizing from all corners

### Screenshots
screen-capture (5)

### CODE REVIEW

File: src/main/components/uml-element/resizable/resizable.tsx [hunk 1]
```diff
@@ -60,39 +91,92 @@ export const resizable =
...
       };
 
-      private onPointerDown = (event: React.PointerEvent<SVGElement>) => {
+      private onPointerDown = (event: React.PointerEvent<SVGElement>, resizeFrom: string) => {
         if (event.nativeEvent.which && event.nativeEvent.which !== 1) {
           return;
         }
 
-        this.setState({ resizing: true, offset: new Point(event.clientX, event.clientY) });
+        switch (resizeFrom) {
```
Comment (martindevtum) [inline src/main/components/uml-element/resizable/resizable.tsx:134]: here i would create a `let offset` which we then assign in the switch statement.
Thus we can have one `this.setState({ resizing: true, offset })` in the end.

Comment (martindevtum) [reply src/main/components/uml-element/resizable/resizable.tsx:134]: exactly as you did in the onPointerMove function.

**Comment by martindevtum** [comment]
Code looks good! Way nicer.

File: src/main/components/uml-element/resizable/resizable.tsx (hunk 1 above)
Comment (minrows) [reply src/main/components/uml-element/resizable/resizable.tsx:135]: @martindevtum 
Done.

**Comment by martindevtum** [comment]
Still one more bug:

**Currently:**
For Elements that can only be scaled in their Y-Axis, e.g. Petri Net Transitions, those elements get repositioned when resized from the Top Left and Bottom Left Corners. 

**Should Be:**
Equivalently to the Bottom Right and Top Right Corners, an element should not be repositioned when they are being resized.


Other than that it looks and feels great.

**Review by martindevtum — CHANGES_REQUESTED** [review]
fix remaining bug.

File: src/main/services/uml-element/uml-element.ts [hunk 2]
```diff
@@ -73,6 +73,7 @@ export abstract class UMLElement implements IUMLElement, ILayoutable {
   strokeColor?: string;
   textColor?: string;
   assessmentNote?: string;
+  resizeFrom = '';
```
Comment (martindevtum) [inline src/main/services/uml-element/uml-element.ts:76]: I would propose to create an enum for this containing the four elements 
TOP_RIGHT,
TOP_LEFT,
BOTTOM_RIGHT,
BOTTOM_LEFT.

It is a bit more descriptive and overall better pratice.

File: src/main/services/uml-element/resizable/resizing-types.ts [hunk 3]
```diff
@@ -12,6 +12,7 @@ export type ResizingActions = ResizeAction;
 export type ResizeAction = Action<ResizingActionTypes.RESIZE> & {
   payload: {
     ids: string[];
+    resizeFrom: string;
```
Comment (martindevtum) [inline src/main/services/uml-element/resizable/resizing-types.ts:15]: Should be an enum, see my other comment.

File: src/main/services/uml-element/resizable/resizing-reducer.ts [hunk 4]
```diff
@@ -1,8 +1,37 @@
 import { Reducer } from 'redux';
+import { IBoundary } from '../../../utils/geometry/boundary';
 import { Actions } from '../../actions';
 import { ResizingActionTypes, ResizingState } from './resizing-types';
 
 export const ResizingReducer: Reducer<ResizingState, Actions> = (state = {}, action) => {
+  const obj = { x: 0, y: 0 };
+
+  const getUpdatedPosition = (elem: IBoundary, payload: any, resizeFrom: string) => {
```
Comment (martindevtum) [inline src/main/services/uml-element/resizable/resizing-reducer.ts:9]: we need to restrict the new positions to not be smaller or greater than the opposing corners via Math.max or Math.min.

Sent you the specifics.

File: src/main/components/uml-element/resizable/resizable.tsx [hunk 5]
```diff
@@ -60,39 +91,94 @@ export const resizable =
...
+            />
           </WrappedComponent>
         );
       }
 
-      private resize = (width: number, height: number) => {
-        width = Math.round(width / 20) * 20;
-        height = Math.round(height / 20) * 20;
+      private resize = (width: number, height: number, resizeFrom: string) => {
+        width = Math.round(width / 5) * 5;
```
Comment (martindevtum) [inline src/main/components/uml-element/resizable/resizable.tsx:119]: I'd think about if we should reset it to 20, now that we found another solution for the repositioning issue.

**Review by martindevtum — CHANGES_REQUESTED** [review]

**Comment by martindevtum** [comment]
feature works as expected.
I approve of the code changes.
👍

**Review by martindevtum — APPROVED** [review]

**Review by bassner — APPROVED** [review]
I've tested this manually and it works fine. Code looks good to me 👍

**Pull request merged**